### PR TITLE
[RPD-164] Check whether remote state bucket has been provisioned

### PR DESCRIPTION
**Pull request by d-lowl**
Check whether:
* State configuration is written and valid
* State bucket is provisioned

## Checklist

Please ensure you have done the following:

* [x] I have read the [CONTRIBUTING](https://github.com/fuzzylabs/matcha/blob/main/CONTRIBUTING.md) guide.
* [ ] I have updated the documentation if required.
* [x] I have added tests which cover my changes.

## Type of change

Tick all those that apply:

* [ ] Bug Fix (non-breaking change, fixing an issue)
* [x] New feature (non-breaking change to add functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Other (add details above)

### CODE REVIEW

File: src/matcha_ml/state/remote_state_manager.py [hunk 1]
```diff
@@ -37,11 +39,23 @@ class RemoteStateManager:
...
 
     def __init__(self, config_path: Optional[str] = None) -> None:
         """Initialise Remote State Manager."""
-        ...
+        if config_path is not None:
+            self.config_path = config_path
+        else:
+            self.config_path = os.path.join(os.getcwd(), DEFAULT_CONFIG_NAME)
+
+    def _configuration_file_exists(self) -> bool:
```
Comment (dudeperf3ct) [inline src/matcha_ml/state/remote_state_manager.py:53]: Docstring missing

File: src/matcha_ml/state/remote_state_manager.py [hunk 2]
```diff
@@ -37,11 +39,23 @@ class RemoteStateManager:
...
+        return os.path.exists(self.config_path)
+
+    def _load_configuration(self) -> RemoteStateConfig:
```
Comment (dudeperf3ct) [inline src/matcha_ml/state/remote_state_manager.py:56]: Docstring missing

File: src/matcha_ml/state/remote_state_manager.py [hunk 3]
```diff
@@ -50,28 +64,49 @@ def configuration(self) -> RemoteStateConfig:
         Returns:
```
Comment (dudeperf3ct) [inline src/matcha_ml/state/remote_state_manager.py:64]: Missing docstring for `Raises`

File: src/matcha_ml/state/remote_state_manager.py [hunk 4]
```diff
@@ -50,28 +64,49 @@ def configuration(self) -> RemoteStateConfig:
...
+        try:
+            return self._load_configuration()
+        except Exception as e:
+            raise MatchaError(f"Error while loading state configuration: {e}")
 
     @property
     def azure_storage(self) -> AzureStorage:
         """Azure Storage property.
 
+        If it was not initialized before, it will be initialized
```
Comment (dudeperf3ct) [inline src/matcha_ml/state/remote_state_manager.py:76]: Missing docstring for `Raises`

File: src/matcha_ml/storage/azure_storage.py
Comment (dudeperf3ct) [inline src/matcha_ml/storage/azure_storage.py:1]: The implementation for `AzureStorage` will change as per PR #101

**Review by dudeperf3ct — CHANGES_REQUESTED** [review]
Excellent @d-lowl. Few missing docstring but implementation looks solid.

**Review by dudeperf3ct — APPROVED** [review]

File: src/matcha_ml/cli/cli.py
Comment (osw282) [inline src/matcha_ml/cli/cli.py:1]: I thought we also need to check if the remote state storage is provisioned in the `provision` command too.

File: src/matcha_ml/state/remote_state_manager.py [hunk 5]
```diff
@@ -37,41 +39,98 @@ class RemoteStateManager:
...
     """
 
+    _azure_storage: Optional[AzureStorage] = None
+
     config_path: str
 
     def __init__(self, config_path: Optional[str] = None) -> None:
         """Initialise Remote State Manager."""
-        ...
+        if config_path is not None:
```
Comment (osw282) [inline src/matcha_ml/state/remote_state_manager.py:48]: Missing args in doc string.

**Review by osw282 — CHANGES_REQUESTED** [review]
One question and a missing arg in doc string, looks good otherwise.

File: src/matcha_ml/cli/cli.py
Comment (d-lowl) [reply src/matcha_ml/cli/cli.py:1]: We're checking the bucket's existence explicitly. I'm not sure if we actually need to check if terraform apply has ever been run.

File: src/matcha_ml/state/remote_state_manager.py [hunk 6]
```diff
@@ -1,10 +1,12 @@
 """Remote state manager module."""
 import contextlib
 import dataclasses
+import os.path
```
[the remark this answers is not in the document]
Comment (d-lowl) [reply src/matcha_ml/state/remote_state_manager.py:4]: Used in __init__

File: src/matcha_ml/state/remote_state_manager.py (hunk 5 above)
Comment (d-lowl) [reply src/matcha_ml/state/remote_state_manager.py:48]: Added in :+1:

**Review by osw282 — APPROVED** [review]
Well done, LGTM!! 🚀

**Pull request merged**